### ISSUE TITLE
"any" handle with path "/*" doesn't run before next handle

### ISSUE THREAD

**Issue by d4n5h** [uNetworking/uWebSockets.js#337]
This does work:
```javascript
const uWS = require('uWebSockets.js');

const app = uWS.App();

app.get('/api', (res, req) => {
    console.log('First')
    req.setYield(true);
    res.onAborted(() => {
        res.aborted = true;
    });
})

app.get('/api',(res,req)=>{
    console.log('Second')
    res.end('Hello');
})

app.listen(9001, (token) => {
    if (token) {
        console.log('Running');
    } else {
        console.log('Failed to run');
    }
})
```
Output:
```
First
Second
```
While this doesn't work:
```javascript
const uWS = require('uWebSockets.js');

const app = uWS.App();

app.any('/*', (res, req) => {
    console.log('First')
    req.setYield(true);
    res.onAborted(() => {
        res.aborted = true;
    });
})

app.get('/api',(res,req)=>{
    console.log('Second')
    res.end('Hello');
})

app.listen(9001, (token) => {
    if (token) {
        console.log('Running');
    } else {
        console.log('Failed to run');
    }
})
```
Output:
```
Second
First
First
First
```

If I comment out `req.setYield(true);` then it will output:
```
Second
First
```

**Comment by hst-m** [comment]
I tested this a bit just now, looks like `app.any()` always matches after the other types like `app.get()`, it doesn't matter if you assign it before the other ones in the JS. The best way for this to work would be to match in the order it was assigned in the JS. In my apps I always use a single `app.any()` to handle all requests so I don't have this issue but seems like it should be fixed

**Comment by hst-m** [comment]
More discussion about this here https://github.com/uNetworking/uWebSockets/issues/975

My recommendation is use a single `app.any(/*)` because then you can code exactly how you want it without worrying about all these routing rules, but maybe it's just my preference

**Comment by alexhultman** [comment]
It's a toipic in the user manual, you can read it there: https://github.com/uNetworking/uWebSockets/blob/master/misc/READMORE.md#pattern-matching

The router is a tree, not a linear vector, so there is no way to have it by order like Express does it. A tree is much more efficient, but has no sane ordering.

**Comment by alexhultman** [comment]
@hst-m Your recommendation routes in O(n), the router in uWS routes in O(log n) if your routes are decently balanced. Maybe this doesn't matter for your use case, but it's inferior, algorithmically.

**Comment by hst-m** [comment]
There are performance considerations and there are logic/flow/readability considerations, @d4n5h is trying to do what every user will want to do which is do some functions before the other routes get called, but currently `any('/*)` won't get called before so it doesn't work, so then instead you have to manually add these functions to every `get()` route which is redundant and difficult to maintain with many routes. Instead if you do a single `any('/*)` you can easily code exactly what you need one time. Regarding performance, if you have critical paths you can optimize your JS to route match those paths first, therefor I believe a single `any('/*)` is good option

**Comment by alexhultman** [comment]
Whatever floats your boat, of course. I don't like the recommendation though. I don't want to promote usage that ignores the router that is already built-in. There's nothing that "cannot be done" or is "hard to maintain" if you know how to compose basic  functions. Even things like "middleware" is really just function composition.

The problem with encouraging people to do their own routing is you end up with projects like Express, where 40% of the overall performance is lost in routing. Your particular snippet is probably fine, but Express is insane - they do RegEx kind of search in linear fashion which severely limits performance and I've worked for companies I cannot legally disclose - but I know cases where the routing list can cut performance down 90%. And that is sad, even if it might now affect the business monetary flow, it still sucks.

So yeah, I have pretty strong opinions here.

**Comment by hst-m** [comment]
I noticed people building things on top of this library doing strange things, I've got app with complex stuff going and everything working well with this library so no complaints so far

**Comment by alexhultman** [comment]
Well, sometimes I just want to buy a farm and grow potatoes for a living.

**Comment by d4n5h** [comment]
@alexhultman @hst-m 
Got it. it makes total sense. I don't want to make the same mistakes as Express, but there are some concepts which I still fond of and often there's a way to achieve the same (or similar) results in a way which doesn't sacrifice performance (too much).
I totally respect and agree with your opinion. This is not Express and for a good reason. That's why I'm trying to move away from Express. Really nice work by the way.

Anyway, I found a workaround that that will solve my problem.
I just automatically generate more routes matching (without regex, just a simple split) the path and method of the child-routes and it works in the same way as a "middleware". works like a charm and barely any performance loss.

**Comment by hst-m** [comment]
you should post here so other people can get ideas on more ways to do it. I am fan of doing it all in JS (but no monolithic/express library on top, minimal amount only)

**Comment by d4n5h** [comment]
Module:
```
const { flatten } = require('array-flatten'),
    uWS = require('uWebSockets.js'),
    fs = require('fs'),
    p = require('path'),
    slice = Array.prototype.slice;

function parseQuery (query) {
    query = query.split('&');
    let results = {};
    for (let i = 0; i < query.length; i++) {
        const kv = query[i].split('=');
        if (kv.length > 1) {
            results[kv[0]] = kv[1];
        }
    }
    return results;
}

const methods = ['get', 'post', 'put', 'del', 'any', 'ws', 'patch', 'listen', 'connect', 'options', 'trace', 'publish', 'head']
const exclude = ['ws', 'listen', 'connect', 'options', 'trace']

const uApp = function (options = {}) {
    this.uWS = uWS;
    if (options.ssl) {
        this.app = this.uWS.SSLApp(options.ssl);
    } else {
        this.app = this.uWS.App();
    }

    this.stack = [];
    this.req = {};

    this.set = function (key, val) {
        this.req[key] = val;
        return this;
    }

    this.use = function use(fn) {
        let offset = 0;
        let path = '/';

        if (typeof fn !== 'function') {
            let arg = fn;

            while (Array.isArray(arg) && arg.length !== 0) {
                arg = arg[0];
            }

            if (typeof arg !== 'function') {
                offset = 1;
                path = fn;
            }
        }

        const callbacks = flatten(slice.call(arguments, offset));

        callbacks.forEach((callback) => {
            if (callback.stack) {
                req = Object.assign(callback.req, this.req)
                callback.stack.forEach((cb) => {
                    this.stack.push({
                        path: path + cb.path, isMw: cb.isMw, method: cb.method, callback: function (res, req) {
                            req = Object.assign(req, this.req)
                            req.query = parseQuery(req.getQuery());
                            cb.callback(res, req)
                        }
                    })
                })
                const newReqs = Object.keys(callback.req);
                for (let i = 0; i < newReqs.length; i++) {
                    const key = newReqs[i];
                    const val = callback.req[key];
                    this.req[key] = val;
                }
            } else {
                if (path == '/') {
                    path = '/*'
                }
                this.stack.push({
                    path: path, isMw: true, method: 'any', callback: (res, req) => {
                        req = Object.assign(req, this.req)
                        req.query = parseQuery(req.getQuery());
                        req.setYield(true);
                        res.onAborted(() => {
                            res.aborted = true;
                        });
                        callback(res, req);
                    }
                })
            }
        })
        return this;
    };

    const that = this;
    methods.forEach(method => {
        uExpress.prototype[method] = function (path, callback) {
            this.stack.push({
                path: path, method: method, callback: function (res, req) {
                    req = Object.assign(req, that.req)
                    req.query = parseQuery(req.getQuery());
                    callback(res, req)
                }
            })
            return this
        }
    });

    this.listen = function (port, cb) {
        if (this.req.static) {
            this.get(this.req.static, (res, req) => {
                const filename = p.join(__dirname, req.getUrl())
                const totalSize = fs.statSync(filename).size;
                const readStream = fs.createReadStream(filename);
                pipeStreamOverResponse(res, readStream, totalSize);
                res.onAborted(() => {
                    res.aborted = true;
                });
            })
        }

        for (let i = 0; i < this.stack.length; i++) {
            const route = this.stack[i];
            if (route.path.includes('*') && route.isMw) {
                const base = route.path.split('*')[0];
                for (let x = 0; x < this.stack.length; x++) {
                    if (this.stack[x].path.includes(base) && !this.stack[x].path.includes('*') && !this.stack[x].isMw && !exclude.includes(this.stack[x].method)) {
                        this.app[this.stack[x].method](this.stack[x].path, route.callback)
                    }
                }
            } else {
                this.app[route.method](route.path, route.callback)
            }
        }
        this.app.listen(port, cb);

    }
    return this;

}

module.exports = uApp;
```

Usage:
```
let app = require('./uApp.js');
const port = 9001;

app = new app();

app.use((res, req) => {
    app.req.test = 'test'
})

app.get('/api/:id', (res, req) => {
    console.log(req.test)
    res.end('asd')
})


app.listen(port, (token) => {
    if (token) {
        console.log('Listening to port ' + port);
    } else {
        console.log('Failed to listen to port ' + port);
    }
})
```

**Comment by hst-m** [comment]
that is a pretty large amount of complex code added on top, plus external library that needs to be installed. In my method of doing this, I would do the same thing in only a couple lines of code instead of 150. Here is how I would do the same example:

```javascript
app.any('/*', (res, req) => {
  const method = req.getMethod()
  const path = req.getUrl()
  res.test = 'test'
  switch(method) {
    case 'get':
      if(path.startsWith('/api/')) {
        const id = path.slice(5)
        console.log(res.test, id)
        return res.end('asd')
      }
  }
  res.writeStatus('404').end()
})
```
then you continue adding endpoints in the same way, no extra large amount of opaque code needed, the least amount of code possible in the clearest and shortest way. You've basically got an Express thing going on there lol was not expecting that. It's not even the performance that is an issue for me with that, it's the added complexity that is just not needed. this is all my opinion however obviously and how I prefer to code

**Comment by d4n5h** [comment]
@hst-m That's also an option. But it doesn't let you pass the additional req variables/objects to routes in other files in recursion.
I don't mind it being similar to Express in a sense of organization. If I could do things in a similar way without sacrificing performance then it's not a bad thing, even with the added complexity. That's why you've got frameworks. And this (in comparison to Express) is extremely streamlined.
Anyway, I can remove the external libraries so it will only be depended on uWS.

**Comment by hst-m** [comment]
I matched your example to show how I would do the same thing, but if you have a different more involved example that showcases more of the code I would be interested in seeing that and matching it in my style to compare, use the most complete and advanced example you can and I could try to match it

**Comment by d4n5h** [comment]
@hst-m 
https://gist.github.com/d4n5h/dbd7f64c7fc09abf9b4eec99401833d5 Try this one and go to: http://localhost:9001/routerPath/api
It will load the routes from route.js to the main stack and pass req to all of them

**Comment by d4n5h** [comment]
@hst-m The benefit of this functionality, for example: let's say you want to check Auth only in one place and pass the result to all routes in different files. Or you want to work on a part of the API in an isolated way. Or reuse a part of the API in another project or even be a standalone app without almost any adjustment to the code. Very much like a separation of microservices.

**Comment by hst-m** [comment]
ok just getting back and looking at this.

Ok so the only difference here is you moved part of this to an external file, and your example you described using this for would be to **_"check Auth only in one place and pass the result to all routes"_**. But that would be very easy to do without this "Route" feature thing and without the external file. In my app I have thousands of subdomains that change programmatically each with multiple types of cookies for multiple users on each subdomain, and I have no problem checking auth in multiple different places and passing those to the different routes, all with fairly simple code (like a couple lines) so I dont see the benefit in this case. 

the 2nd example you say would be to **_"reuse a part of the API in another project or even be a standalone app without almost any adjustment to the code"_**. For my server app I only need 1 file so this applies less to me, so it's possible this is helpful to you in some way, though I would think you could do it in less code, but it would depend on what you are trying to do. Some side trivia: Deno (Node 2.0) is now refactoring their entire codebase into a single 10k line file because it is easier to work with than many individual files (see dev [link](https://docs.google.com/document/d/1_WvwHl7BXUPmoiSeD8G83JmS8ypsTPqed4Btkqkn_-4/preview?pru=AAABcrrKL5k*nQ4LS569NsRRAce2BVanXw#))

so in conclusion I still believe it is important to use as little code as possible, you need to save the complexity for parts that really need it and keep everything else as simple as possible. this is just what works for me though and I enjoy looking at what other people are doing

**Comment by d4n5h** [comment]
@hst-m There is no silver bullet which will fit in all circumstances. I understand your perspective, and for you it make sense to use this library as-is. I really don't know what is complex in what I did, with my code you could still use uWS in the exact same way but you'll have a few other functions that can make some workflows a bit more easier and it shouldn't effect performance since the routes are converted to regular uWS routes at build time and there's no change in the way the route system works.

**Comment by hst-m** [comment]
I asked for an updated example that added functionality, and all the updated code does is add a 2nd basic endpoint `/api` except now it spans more files than the 1st example and still uses all of that code that is completely unnecessary, so I have still not seen an example of where any of this is necessary, maybe I am missing something

**Comment by d4n5h** [comment]
What you don't understand?... If you want to separate the router/app to multiple files because for most people (clearly not you) it's easier to maintain especially if you work in a team which each person work on different part of the app and you want to test each part separately in isolation or to reuse it and a different app. If you don't see the value in it but you say that "but no monolithic..." then it's a bit strange.  Anyway, no. I will not work on a 10k line file and I will not use this library in the way that you suggested (creating one any route and using a switch) lol. Maybe I'm weird but when I open a browser I like to have multiple tabs and not one tab with the content of 20 websites. And I prefer to read a book with muiltple pages and not a scroll. Need more examples?

**Comment by d4n5h** [comment]
@hst-m anyway, could you provide me with an example that achieves the same result (separate router/app files with shared req objects) without using your solution of using one any wildcard route and a switch?

**Comment by hst-m** [comment]
My app is only 200 lines of code so far, obviously 10k is for a major application like Deno. It is true I am the only developer so it is much easier for me to code only what is needed. I previously worked with a team for a company and it was very difficult to work on things in a way everyone agrees. In fact it tore the company apart and resulted in massive amount of bugs and the product never launched because the entire thing was way over-complicated between the front-end and back-end, that is where I get my philosophy from. I can look into doing what you requested

**Comment by d4n5h** [comment]
> I asked for an updated example that added functionality, and all the updated code does is add a 2nd basic endpoint `/api` except now it spans more files than the 1st example and still uses all of that code that is completely unnecessary, so I have still not seen an example of where any of this is necessary, maybe I am missing something 

I will not write you a whole api just for an example. I think that you can imagine how this example might scale.

**Comment by d4n5h** [comment]
> My app is only 200 lines of code so far (client-side at 650 vanilla js for the main SPA page), obviously 10k is for a major application like Deno. It is true I am the only developer so it is much easier for me to code only what is needed. I previously worked with a team for a company and it was very difficult to work on things in a way everyone agrees. In fact it tore the company apart and resulted in massive amount of bugs and the product never launched because the entire thing was way over-complicated between the front-end and back-end, that is where I get my philosophy from. I can look into doing what you requested

Seems like bad project management. You should've use scrum/kanban. By the way, another great reason why to use frameworks.

**Comment by hst-m** [comment]
Yes it is 100% dependent on whoever is in charge, and the lead devs were clueless and there was nothing I could do because I disagreed with everything they were doing. I am very anti-framework though and enjoy my vanilla javascript, and is why I like this uws library because it is simple/basic does what it needs and thats it

**Comment by alexhultman** [comment]
> Deno (Node 2.0) is now refactoring their entire codebase into a single 10k line file because it is easier to work with than many individual files (see dev link)

Monkeys throw their feces. That doesn't mean we should copy them. The Pope rapes children. That doesn't mean we should.

Deno is written by amateur hipsters with no direction or leadership. Anything goes, as long as you use correct pronouns and only bring positive feedback. It's a safe space community. They are insane.

**Comment by hst-m** [comment]
well you cut out where I said "side trivia" because it is mostly meaningless, mostly entertaining to see how other people are doing things, their internal debate documents leaked, and happens to be somewhat related in that its supposed to be http server etc

**Comment by alexhultman** [comment]
I don't think they have anything internal. There's no leadership. Ryan is a people pleaser, he'll say yes to everyone then sail in randomly changing direction.

I told them from the start before they even began that they have to implement as much as possible in native code (99%) and he says yes, then does the complete opposite and gets shit poor performance. Then I tell him the performance is crap and I get banned.

**Comment by alexhultman** [comment]
@hst-m That document you linked doesn't even say they moved 10k lines in one file for ease of use, they did that because they had "bootstrapping problems" and TypeScript problems. Linking to something you haven't even read is very lazy.

**Comment by hst-m** [comment]
very clearly says **_"this will allow new contributors to quickly understand how Deno bootstraps itself"_** which means it is easier to use and understand, which I tend to agree with in my own experience

**Comment by hst-m** [comment]
then it says **_"Build-time complexity and performance is obviously better with this setup"_**

**Comment by alexhultman** [comment]
You're focusing on select sentences and pulling them out of context. The context is that TypeScript enforces certain parts they dislike. That's their problem definition.


Look I can also pull individual sentences out of context:

> the idea of having a 10k line javascript file sounds unappealing

**Comment by alexhultman** [comment]
Thanks for today, I'll be removing this issue
